Add Climber

dslit3 opened the pull request:
Adds climbers
Climbers zero on teleop start and can be re-zeroed with a controller button

Also renames `MoPrefs#amperesPref` to `MoPrefs#ampsPref` because I made it before it was added to main separately and I kept the one that matched WPILib's unit name

dslit3 committed:
Add climbers

Code is not tested
No CAN addresses yet
Prefs will need to be changed

dslit3 committed:
Merge branch 'main' of https://github.com/momentumfrc/2024Crescendo into climber

dslit3 committed:
Add climber shuffleboard tab

dslit3 committed:
Spotless

dslit3 committed:
Merge branch 'main' of https://github.com/momentumfrc/2024Crescendo into climber

dslit3 committed:
Merge main

jordan-powers commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -50,11 +54,14 @@ public class RobotContainer {
     private TeleopDriveCommand driveCommand = new TeleopDriveCommand(drive, positioning, this::getInput);
     private TeleopArmCommand armCommand = new TeleopArmCommand(arm, this::getInput);
+    private SequentialCommandGroup climbCommand =
+            new SequentialCommandGroup(new ZeroClimbersCommand(climb), new ClimbCommand(climb, this::getInput));
```
If the climbers are already zeroed, we don't need to re-zero every time the robot is enabled. You should make the default command just the `ClimbCommand`, then use a trigger to re-zero on Autonomous or Teleop start, but only if the climber is not already zeroed.

Basically, duplicate the logic that already exists for the intake. See line 182 in this file.

jordan-powers commented on src/main/java/frc/robot/command/climb/ZeroClimbersCommand.java:
```diff
@@ -0,0 +1,30 @@
+import frc.robot.subsystem.ClimbSubsystem;
+
+public class ZeroClimbersCommand extends Command {
+    private static final double POWER = 0.3;
```
This should be a MoPref.

```java
public static final UnitPref<Dimensionless> climberZeroPower = getInstance().new UnitPref<Dimensionless>("Climber Zero Power", Units.Percent, Units.Percent.of(30));


double power = MoPrefs.climberZeroPower.get().in(Units.Value);
```

jordan-powers commented on src/main/java/frc/robot/component/Climber.java:
```diff
@@ -0,0 +1,99 @@
+    public Climber(
+            String name, CANSparkMax winch, UnitPref<Current> zeroCurrentLimit, UnitPref<Time> zeroTriggerDuration) {
+        this.winch = winch;
+        this.winchPID = winch.getPIDController();
```
No, we don't need pid on this. Straight motor power should be perfectly fine, and we don't have time to waste tuning unnecessary pid loops.

jordan-powers commented on src/main/java/frc/robot/component/Climber.java:
```diff
@@ -0,0 +1,99 @@
+    private boolean hasZero = false;
+
+    public Climber(
+            String name, CANSparkMax winch, UnitPref<Current> zeroCurrentLimit, UnitPref<Time> zeroTriggerDuration) {
```
You're missing the standard motor configuration preamble. Refer to the first few lines of the IntakeSubsystem and ArmSubsystem constructors.

Basically, you need to configure:
- Smart current limit
- Idle mode
- Soft limits

jordan-powers commented on src/main/java/frc/robot/component/Climber.java:
```diff
@@ -0,0 +1,99 @@
+            return;
+        }
+
+        if (power > 0 && winch.getEncoder().getPosition() >= MoPrefs.climberZeroThreshold.get()) {
+            this.runWinch(0);
+            return;
+        }
+
```
We don't use raw encoders anymore. All encoders should be wrapped in a MoEncoder instance. That way, you don't have to think about the units of the encoder. The MoEncoder encapsulates any unit conversion, and just returns a Measure of the current encoder value.

 Refer the IntakeSubsystem and ArmSubsystem for examples.

jordan-powers commented on src/main/java/frc/robot/component/Climber.java:
```diff
@@ -0,0 +1,99 @@
+            return;
+        }
+
+        if (power > 0 && winch.getEncoder().getPosition() >= MoPrefs.climberZeroThreshold.get()) {
```
1. We should be using soft limits on the motor controller, so you shouldn't need to check extents yourself. Just make sure the soft limits are set and enabled in the ClimberSubsystem constructor. (Although make sure to disable the reverse soft limit when re-zeroing.)
2. You're using the same climberZeroThreshold for both the forward and reverse direction? Wouldn't that mean that the effective range of travel is 0, and the climber will never move? Usually, you set the reverse direction limit to 0, and the forward direction limit to the value from MoPrefs.

jordan-powers commented on src/main/java/frc/robot/component/Climber.java:
```diff
@@ -0,0 +1,99 @@
+        this.runWinch(power);
+    }
+
+    public void zero(double power) {
+        if (!hasZero) {
+            winch.set(-power);
+
+            if (atZeroLimit()) {
+                winch.getEncoder().setPosition(0);
+                hasZero = true;
+                winch.set(0);
+            }
+        } else {
+            winch.set(0);
+        }
+    }
```
This logic should be kept in the ZeroClimberCommand, not in the subsystem.

Refer to the ZeroIntakeCommand for an example.

jordan-powers commented on src/main/java/frc/robot/component/CurrentTrigger.java:
```diff
@@ -0,0 +1,57 @@
+    public static CurrentTrigger ofSparkMax(
+            CANSparkMax spark, UnitPref<Current> limitPref, UnitPref<Time> durationPref) {
+        var trigger =
+                new CurrentTrigger(() -> Units.Amps.of(spark.getOutputCurrent()), limitPref.get(), durationPref.get());
```
Don't use `Unit.of` anywhere that will be called in a loop. It creates a new Measure instance every time it is called, which is hell on the garbage collector. And we're already really short on memory this year. Instead, use a `MutableMeasure.mut_replace`.

See https://docs.wpilib.org/en/stable/docs/software/basic-programming/java-units.html#mutability-and-object-creation

jordan-powers commented on src/main/java/frc/robot/component/CurrentTrigger.java:
```diff
@@ -0,0 +1,57 @@
+    @Override
+    public boolean getAsBoolean() {
+        if (current.get().gt(currentThreshold)) {
+            return (System.currentTimeMillis() - triggerStartTimeMs) > triggerDuration.in(Units.Milliseconds);
```
Don't write the timer logic yourself. Use a `edu.wpi.first.wpilibj.Timer` instance.

jordan-powers commented on src/main/java/frc/robot/component/CurrentTrigger.java:
```diff
@@ -0,0 +1,57 @@
+
+    private long triggerStartTimeMs = 0;
+
+    public CurrentTrigger(
```
Not sure why this needs to be broken out into its own class, when the core logic is just a 4-line if-statement.

Refer to the ZeroIntakeCommand to see how I would do it.

jordan-powers commented on src/main/java/frc/robot/input/JoystickDualControllerInput.java:
```diff
@@ -135,4 +135,19 @@ public boolean getSaveIntakeSetpoint() {
         return armController.getAButton();
     }
+
+    @Override
+    public double getLeftClimbRequest() {
+        return (armController.getLeftTriggerAxis() - 0.5) * 2;
+    }
+
+    @Override
+    public double getRightClimbRequest() {
+        return (armController.getRightTriggerAxis() - 0.5) * 2;
+    }
```
This maps the range of each trigger from -1 to 1. This means that when the driver is not pulling on the triggers, the climber will receive an input of -1. This means that the motor will be running at full reverse the whole match unless the driver "feathers" the triggers and holds them exactly in the middle of their range of travel.

This is not an acceptable user experience. The default behavior if the driver is not touching the controller should be for the climbers to stay stationary. Then we want some way to both raise and lower each climber individually.

My suggestion: have each trigger mapped to return a value between 0 and 1, such that when the trigger is not pulled it returns 0 and when the trigger is full pulled it returns 1. Then add another "reverse" button to the controller, so while the button is held, the commanded input is reversed. So if the driver pulls a trigger the climber starts going down, but if the driver also holds the "reverse" button, the climber starts going up.

dslit3 replied:
There are very few available buttons for this, but the newest commit uses pov=90 (Currently unused D-Pad button) as a reverse switch

jordan-powers commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -125,13 +133,15 @@ public RobotContainer() {
         runSysidTrigger = new Trigger(() -> getInput().getRunSysId());
         shootSpeakerTrigger = new Trigger(() -> getInput().getShouldShootSpeaker());
         shootAmpTrigger = new Trigger(() -> getInput().getShouldShootAmp());
+        reZeroClimbTrigger = new Trigger(() -> getInput().getReZeroClimbers());
```
Not sure we need a button to re-zero the climber. Why not just a switch on the shuffleboard? It shouldn't need to happen all that often.

jordan-powers commented on src/main/java/frc/robot/component/Climber.java:
```diff
@@ -0,0 +1,99 @@
+
+    private void runWinch(double power) {
+        if (MoPrefs.climberPid.get()) {
+            winchPID.setReference(
+                    power * MoPrefs.climberMotorSpeed.get().in(MoUnits.RotationsPerMinute),
+                    CANSparkMax.ControlType.kVelocity);
```
Hmm, we shouldn't be using raw SparkPIDControllers either. This should be wrapped in a MoSparkPIDController, which provides two benefits:
1. It allows the use of the TunerUtils to tune the PID constants.
2. It handles the units. So you just pass your Velocity measure in, and it will determine the units necessary to be sent to the PID controller.

Refer to the ArmSubsystem/IntakeSubsystem/SwerveModule classes for examples.

jordan-powers replied:
Although actually we should just take the PID out entirely. It's not necessary in this situation.

jordan-powers commented on src/main/java/frc/robot/util/MoUnits.java:
```diff
@@ -29,6 +29,6 @@ public static class EncoderAngle extends Unit<EncoderAngle> {
     public static Per<EncoderAngle, Angle> EncoderTicksPerRotation = EncoderTicks.per(Units.Rotations);
     public static Per<EncoderAngle, Angle> EncoderTicksPerRadian = EncoderTicks.per(Units.Radians);
-
+    public static Velocity<Angle> RotationsPerMinute = Units.Rotations.per(Units.Minutes);
```
This unit shouldn't be necessary. We keep all our own calculations in rotations per second.

While we do need to format into rotations per minute to pass to the Spark Maxes, that's handled by the MoEncoder's ability to determine the motor controller's "velocity base unit".

jordan-powers commented on src/main/java/frc/robot/subsystem/ClimbSubsystem.java:
```diff
@@ -0,0 +1,40 @@
+        rightClimber =
+                new Climber("Climber Right", rightSparkMax, MoPrefs.climberCurrentThreshold, MoPrefs.climberLimitTime);
+
+        MoShuffleboard.getInstance().settingsTab.add("Invalidate Climber Zeros", new InstantCommand(() -> {
+            leftClimber.invalidateZero();
+            rightClimber.invalidateZero();
+        }));
```
I've found command buttons on the shuffleboard to be unreliable in the past. Better to just keep the boolean itself on the shuffleboard, instead of adding a command to reset some other non-shuffleboard boolean.

Refer to how the IntakeSubsystem keeps track of whether it has been zeroed.

jordan-powers committed:
Merge branch 'main' into climber

dslit3 committed:
Improve climber

- Zero routine no longer happens on every enable
- Remove PID on climber
- Use smart current limit and soft limits on climb winch
- Use distance units for climb encoders
- Climb motors run in brake mode (!!!)
- Climb controls are now similar to 2022

dslit3 committed:
Fix climber distance value on shuffleboard

dslit3 committed:
Merge branch 'main' of https://github.com/momentumfrc/2024Crescendo into climber

dslit3 committed:
Move zero logic and remove/rename stuff

- `ClimbCommand` -> `TeleopClimbCommand`
- Climb zeroing logic is now entirely in the command
- Remove `CurrentTrigger`

jordan-powers commented on src/main/java/frc/robot/command/climb/ZeroClimbersCommand.java:
```diff
@@ -0,0 +1,66 @@
+        climb.rightClimber.enableWinchSoftLimitReverse(false);
+    }
+
+    private static void zeroSide(double power, Climber climber, Timer currentTimer) {
+        if (climber.winch.getOutputCurrent()
+                >= MoPrefs.climberZeroCurrentCutoff.get().in(Units.Amps)) {
+            if (currentTimer.hasElapsed(MoPrefs.climberZeroTimeCutoff.get().in(Units.Seconds))) {
+                climber.encoder.setPosition(Units.Centimeters.of(0));
+                climber.hasZero.setBoolean(true);
+            }
+        } else {
+            currentTimer.restart();
+        }
+
+        climber.runWinch(-Math.abs(MoPrefs.intakeZeroPwr.get()));
+    }
+
+    @Override
+    public void execute() {
+        zeroSide(MoPrefs.climberZeroPwr.get(), climb.leftClimber, leftCurrentTimer);
+        zeroSide(MoPrefs.climberZeroPwr.get(), climb.rightClimber, rightCurrentTimer);
+    }
```
Once one side has a zero, we should not be commanding power to it anymore, even if the other side hasn't zeroed yet. 

This isn't a problem in the ZeroIntakeCommand because once we have a zero, the command exits. But here, one side can zero before the other is finished, but we'll keep commanding `climber.runWinch(-Math.abs(MoPrefs.intakeZeroPwr.get()));` to both sides.

One solution is as follows:
1. In `initialize()`, set `climber.hasZero` to false for both sides.
2. In `zeroSide()`, check if `climber.hasZero` is true. If so, command zero power. Otherwise, command `intakeZeroPwr`.

jordan-powers commented on src/main/java/frc/robot/component/Climber.java:
```diff
@@ -0,0 +1,75 @@
+        this.encoder = MoEncoder.forSparkRelative(winch.getEncoder(), Units.Centimeters);
+
+        MoPrefs.climberEncoderScale.subscribe(this.encoder::setConversionFactor, true);
+        MoPrefs.climberZeroCurrentCutoff.subscribe(
+                current -> winch.setSmartCurrentLimit((int) current.in(Units.Amps)), true);
```
The zeroing current threshold and the smartCurrentLimit should be entirely different values. The zeroing current should be a low threshold that we just use to zero when we're up against the hard stops. But the smartCurrentLimit is a very high threshold that should only trip if we're at risk of damaging the motor.

Think, if we're lifting the whole robot, we'll be using a lot of power and current. So the smartCurrentLimit should be high so the motor can actually provide the requisite power. But we don't want to apply all that power against the hard limits on the climber--we could break something! The zeroing current threshold should be a low value, still greater than the current used by climber when not loaded, but not as much as we'd use when actually climbing.

The zeroing current is a value we'll have to tune, so it should be a robot preference. But the smartCurrentLimit is a value inherent to the motor we're using (40-60A for a Neo, [per rev](https://docs.revrobotics.com/sparkmax/gs-sm/connect-a-spark-max-over-usb#limiting-current)), so it can stay a class constant.

Refer to the ArmSubsystem and IntakeSubsystem for examples.

jordan-powers commented on src/main/java/frc/robot/util/MoPrefs.java:
```diff
@@ -135,14 +139,23 @@ public class MoPrefs {
+            ampsPref("Climber Zero Current", Units.Amps.of(80));
+    public static final UnitPref<Velocity<Angle>> climberMotorSpeed =
+            rotationsPerSecPref("Climber Motor Speed", Units.RotationsPerSecond.of(75));
+    public static final UnitPref<Distance> climberMinimum = centimetersPref("Climber Minimum", Units.Centimeters.of(2));
```
The preference `climberMinimum` isn't used anywhere that I can see.

There are two approaches to ensuring that the climber doesn't actually run against the hard stops when in normal operation.
1. When zeroing, you actually set the encoder to some negative value. Then you set the soft limit to zero.
2. When zeroing, you set the encoder to zero. Then, you set the soft limit to something greater than zero.

Either approach will work. We've been using 1. on all the other subsystems, but feel free to use whichever you prefer.

But do implement one of these approaches. Currently you're both setting the encoder to zero when zeroing, and you're setting the soft limit to zero. This leaves no margin for error if the motor should happen drive a bit past the soft limits between iterations of the control loop. Better to use one of these approaches to add some buffer, so the mechanism never actually contacts the limits (except when zeroing).

jordan-powers commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -172,6 +180,8 @@ private void configureBindings() {
 
         handoffTrigger.and(() -> !tuneSetpointSubscriber.getBoolean(false)).whileTrue(handoffCommand);
 
+        reZeroClimbTrigger.onTrue(new ZeroClimbersCommand(climb));
+
```
Hmm, we'll need to do some testing to see if this actually triggers when the robot first boots.

I'm not 100% certain how the `.onTrue()` decorator works. If it's only on a rising edge, and the trigger starts out at `true`, then there's a possibility it won't trigger.

If you want to be safe, I'd add another RobotModeTriggers trigger, like with the intake.

Something like
```java
(RobotModeTriggers.autonomous().or(RobotModeTriggers.teleop()))
                .and(() -> !climb.bothZeroed())
                .onTrue(zeroClimbCommand);
````

jordan-powers reviewed (changes requested):
Much better! Just a couple points left

momentumrobotics-4999 committed:
Use brushless intake roller motor

momentumrobotics-4999 committed:
Merge branch 'main' into climber

momentumrobotics-4999 committed:
Improve zero logic and fix current limits
- Also changes "rezero" to "reZero" for intake fields to match the rest of the codebase

momentumrobotics-4999 committed:
Invalidate climber zeros before zeroing
